Only allow the Checked C extension flag for C.

dtarditi opened the pull request:
This change only allows the -fcheckedc-extension flag to be used for C programs
in clang.  The clang driver will reject the use of -fcheckedc-extension for
other C family languages supported by clang, including C++, Objective
C, OpenCL, and CUDA. This addresses issue #9 in the checked-clang Github repo.

We are currently not modifying clang to support these other languages, which
is why need to disallow using the extension with them.

Testing:
-  Add 4 new tests to clang.  They test that use of the extension flag is
rejected for C++, Objective C, CUDA, and OpenCL.
- As recommended by the clang documentation, I placed the tests in with
  other similar tests in the test tree.
- I updated the testing baselines to reflect the new tests.  I also updated
the documentation to reflect the fact that we have Checked C specific tests
in clang. We need to take the new tests into account when updating to new
versions of the clang/LLVM sources.

msftclas commented:
Hi __@dtarditi__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (David Tarditi). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

reubeno commented on docs/checkedc/Test-Baselines.md:
```diff
@@ -52,6 +89,16 @@ Here is the current base line for testing LLVM + clang on all targets (check-all
+in these configurations.
+
+- For just clang with the x86 target (using the check-clang project), 3 additional `Expected Passes` tests.
+- For testing LLVM + clang on x86 (check-all), 8 additional `Expected Passes` tests
```
I saw your updated instructions for keeping this .md file up to date, but I just get the sense that it's going to get out-of-sync.  (And if it stays in sync, I'm wondering if it's going to be unnecessary developer burden for each change that adds a new test.)

I know it's important to keep in agreement on which failures are expected (because they show up in the base line) and agree we should keep that updated--but do we need to update the counts of tests we've added?  I'm not sure how much we'll get from that...

dtarditi replied:
Agreed.  Updating precise counts will likely become a developer burden.  We only need to track 'Unexpected failures' in the baseline branch.  We don't particularly need to track the counts. I'll update the md files in the pull request to reflect that. 

I did have some concerns that people will add tests incorrectly and not notice they aren't running.   However, this seems pretty unlikely given the way that the LLVM test infrastructure  works.  It will produce an error for tests that are lacking RUN lines.  The only requirement during code reviews to make sure that tests will run is to make sure they are placed within test suite directories.

reubeno commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1701,7 +1701,25 @@ static void ParseLangArgs(LangOptions &Opts, ArgList &Args, InputKind IK,
+    else if (Opts.CPlusPlus)
+      disallowed = "C++";
+    else if (Opts.ObjC1 || Opts.ObjC2)
+      disallowed = "Objective C";
```
Are these options exclusive?  If not, should we capture the list of all of the conflicting options rather than a single string?

dtarditi replied:
These options are mutually exclusive, except that Objective C code may be compiled side-by-side with C++ code.  CUDA, OpenCL, and Objective C are all treated as different languages by clang (see clang\include\clang\Basic\LangOptions.def for a list of the language options).   I've updated the message for Objective C when compiled side-by-side with C++.  I also switched to using std::string instead of raw C strings, just to clean up the code.

reubeno commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1701,7 +1701,25 @@ static void ParseLangArgs(LangOptions &Opts, ArgList &Args, InputKind IK,
+    else if (Opts.ObjC1 || Opts.ObjC2)
+      disallowed = "Objective C";
+    if (disallowed) {
+      Diags.Report(diag::err_drv_argument_not_allowed_with) <<
```
Will this end up being fatal?  Or just a warning?  Or will it just not enable the Checked C Extensions?

dtarditi replied:
This ends up being fatal.

reubeno commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1701,7 +1701,25 @@ static void ParseLangArgs(LangOptions &Opts, ArgList &Args, InputKind IK,
+        "-fcheckedc-extension" << disallowed;
+    } else if (!(Opts.C99 || Opts.C11)) {
+      Diags.Report(diag::err_drv_argument_only_allowed_with) <<
+        "-fcheckedc-extension" << "C";
```
Should we be more precise and call out that only C99 and C11 are supported?  I don't remember if clang supports ANSI C under a separate option...

dtarditi replied:
This is a bug.   I was trying to protect against someone adding a new major language extension by specifically calling out that only C was supported.  However, I misread the code - there is no way to say we "only support this  compiler flag with plain C" in the current system.   

After studying the code more carefully, I realized that clang assumes that ANSI C (C89) is the default and then builds upon that with various flags for C++ and newer versions of the C and C++ standards. We do want to support  to C89.  The fix is to remove the specific checks for C99 and C11.

For the way that clang represents language options, see clang\include\clang\Basic\LangOptions.def.    This is included by LangOptions.h, in the same directory.

The way clang internally represents language options is different from the default language version that is chosen if none is specified.   The driver currently sets the default language version for C to C11 with gnu extensions.

reubeno commented on test/Driver/checkedc-notsupported.cpp:
```diff
@@ -0,0 +1,9 @@
+// Checked C extension is not supported for C++.   Make sure driver
```
It's probably low priority, but we could also add a test case for a .c file that is forced to be compiled as C++ using command-line options.  That's a pattern I've certainly seen before.

dtarditi replied:
I've added test cases forcing files to be compiled as different file types than the file extension specifies, with the Checked C extension flag enabled:

- I added a new test with a C file type. It is compiled as a regular C file and also as a C++ file..
- I added run lines where several file types are forced to be compiled as C files..
- I added run lines where the cl file type is compiled as several other language types.
- Finally, I also added tests for clang-cl testing the Microsoft command-line options /TP and /TC for compiling files as C++ and C files respectively.

reubeno commented:
Finished taking a pass over this; my main question is around the overhead to maintain the expected test counts for tests we add.  Apart from that, this looks good.  (And thanks for putting in the effort to add those test cases!)

dtarditi commented:
I've updated the documentation so that we will not maintain expected passing test counts, which could add overhead. Instead we'll just make sure that there are no unexpected test failures.   I addressed your other comments and updated the issue9 branch.  One neat thing: we're now checking that clang-cl understands the -fcheckedc-extension flag when it is passed using the -Xclang option to clang-cl.
